plugins/pixart-rf/fu-pxi-device.c

```diff
@@ -0,0 +1,515 @@
+	if (!fu_pxi_device_wait_notify (self, 0x1, &opcode, NULL, error))
+		return FALSE;
+	if (opcode != FU_PXI_DEVICE_CMD_FW_UPGRADE) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_READ,
+			     "FwUpgrade opcode invalid %02x",
+			     opcode);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_pxi_device_write_firmware (FuDevice *device,
+			      FuFirmware *firmware,
+			      FwupdInstallFlags flags,
+			      GError **error)
+{
+	FuPxiDevice *self = FU_PXI_DEVICE (device);
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* get the default image */
+	fw = fu_firmware_get_image_default_bytes (firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	/* send fw ota init command */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_BUSY);
+	if (!fu_pxi_device_fw_ota_init (self, error))
+		return FALSE;
+	if (!fu_pxi_device_fw_ota_init_new (self, g_bytes_get_size (fw), error))
+		return FALSE;
+
+	/* write fw into device */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_WRITE);
+	chunks = fu_chunk_array_new_from_bytes (fw, 0x0, 0x0, FU_PXI_DEVICE_OBJECT_SIZE_MAX);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index (chunks, i);
+		if (!fu_pxi_device_write_chunk (self, chk, error))
+			return FALSE;
+		fu_device_set_progress_full (device, (gsize) i, (gsize) chunks->len);
+	}
+
+	/* fw upgrade command */
+	if (!fu_pxi_device_fw_upgrade (self, firmware, error))
+		return FALSE;
+
+	/* send device reset command */
+	return fu_pxi_device_reset (self, g_bytes_get_size (fw), error);
+}
+
+static gboolean
+fu_pxi_device_fw_get_info (FuPxiDevice *self, GError **error)
+{
+	guint8 req[64] = { 0x0 };
+	guint8 opcode = 0x0;
+	guint16 checksum = 0;
+	g_autofree gchar *version_str = NULL;
+	g_autofree gchar *checksum_str = NULL;
+
+	req[0] = PXI_HID_DEV_OTA_OUTPUT_REPORT_ID;
+	req[1] = FU_PXI_DEVICE_CMD_FW_GET_INFO;
+	if (!fu_udev_device_pwrite_full (FU_UDEV_DEVICE (self), 0,
+					 (guint8 *) req, 2, error))
+		return FALSE;
+
+	req[0] = PXI_HID_DEV_OTA_FEATURE_REPORT_ID;
+	req[1] = FU_PXI_DEVICE_CMD_FW_GET_INFO;
+	if (!fu_pxi_device_get_feature (self, req, sizeof(req), error))
+		return FALSE;
+	if (g_getenv ("FWUPD_PIXART_RF_VERBOSE") != NULL)
+		fu_common_dump_raw (G_LOG_DOMAIN, "req", (guint8 *) req, sizeof(req));
+	if (!fu_common_read_uint8_safe (req, sizeof(req), 0x2, &opcode, error))
+		return FALSE;
+	if (opcode != FU_PXI_DEVICE_CMD_FW_GET_INFO) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "FwGetInfo opcode invalid %02x",
+			     opcode);
+		return FALSE;
+	}
+
+	/* set current version */
+	version_str = g_strndup ((gchar *) req + 0x3, 5);
+	fu_device_set_version (FU_DEVICE (self), version_str);
+
+	/* add current checksum */
+	if (!fu_common_read_uint16_safe (req, sizeof(req), 0x8,
+					 &checksum, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	checksum_str = g_strdup_printf ("0x%4x", checksum);
```

Comment:
@jyu99 so this is just a uint16_t byte sum, right? I don't know if this is "good enough" to use for verification. @superm1 what do you think?

Reply:
Yes,our checksum is only unit16_byte

Reply:
our checksum is most

Reply:
I've removed it, it'll make the security team nervous...